File: project_management/project_management/report/task_progress_timeline/task_progress_timeline.py
# Copyright (c) 2025, Aman and Yashvi and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import today, add_days, getdate, get_datetime, date_diff, now_datetime
from frappe.query_builder.functions import Count, Sum, Avg
from datetime import datetime, timedelta
import json

def execute(filters=None):
    if not filters:
        filters = {}
    
    if not filters.get("time_interval"):
        filters["time_interval"] = "this_week"
    
    date_range = get_date_range(filters)
    filters["from_date"] = date_range["from_date"]
    filters["to_date"] = date_range["to_date"]
    
    data = get_report_data(filters)
    columns = get_columns(filters)
    chart = get_task_status_chart(data)
    summary = get_report_summary(data, filters)
    
    return columns, data, None, chart, summary

def get_date_range(filters):
    today_date = getdate(today())
    
    if filters.get("from_date") and filters.get("to_date"):
        return {
            "from_date": filters.get("from_date"),
            "to_date": filters.get("to_date")
        }
    
    time_interval = filters.get("time_interval")
    
    if time_interval == "today":
        return {
            "from_date": today_date,
            "to_date": today_date
        }
    elif time_interval == "yesterday":
        yesterday = add_days(today_date, -1)
        return {
            "from_date": yesterday,
            "to_date": yesterday
        }
    elif time_interval == "this_week":
        this_week_start = add_days(today_date, -today_date.weekday())
        this_week_end = add_days(this_week_start, 6)
        return {
            "from_date": this_week_start,
            "to_date": this_week_end if this_week_end <= today_date else today_date
        }
    elif time_interval == "last_week":
        last_week_start = add_days(today_date, -today_date.weekday() - 7)
        last_week_end = add_days(last_week_start, 6)
        return {
            "from_date": last_week_start,
            "to_date": last_week_end
        }
    elif time_interval == "this_month":
        month_start = today_date.replace(day=1)
        return {
            "from_date": month_start,
            "to_date": today_date
        }
    elif time_interval == "last_month":
        current_month_start = today_date.replace(day=1)
        last_month_end = add_days(current_month_start, -1)
        last_month_start = last_month_end.replace(day=1)
        return {
            "from_date": last_month_start,
            "to_date": last_month_end
        }
    elif time_interval == "this_quarter":
        quarter = (today_date.month - 1) // 3 + 1
        quarter_start_month = (quarter - 1) * 3 + 1
        quarter_start = today_date.replace(month=quarter_start_month, day=1)
        return {
            "from_date": quarter_start,
            "to_date": today_date
        }
    elif time_interval == "last_quarter":
        current_quarter = (today_date.month - 1) // 3 + 1
        last_quarter = current_quarter - 1 if current_quarter > 1 else 4
        last_quarter_year = today_date.year if current_quarter > 1 else today_date.year - 1
        last_quarter_start_month = (last_quarter - 1) * 3 + 1
        last_quarter_start = datetime(last_quarter_year, last_quarter_start_month, 1).date()
        last_quarter_end_month = last_quarter * 3
        if last_quarter_end_month == 12:
            last_quarter_end = datetime(last_quarter_year, 12, 31).date()
        else:
            last_quarter_end = datetime(last_quarter_year, last_quarter_end_month + 1, 1).date()
            last_quarter_end = add_days(last_quarter_end, -1)
        return {
            "from_date": last_quarter_start,
            "to_date": last_quarter_end
        }
    elif time_interval == "this_year":
        year_start = today_date.replace(month=1, day=1)
        return {
            "from_date": year_start,
            "to_date": today_date
        }
    elif time_interval == "last_year":
        last_year_start = datetime(today_date.year - 1, 1, 1).date()
        last_year_end = datetime(today_date.year - 1, 12, 31).date()
        return {
            "from_date": last_year_start,
            "to_date": last_year_end
        }
    elif time_interval == "all":
        return {
            "from_date": datetime(2000, 1, 1).date(),
            "to_date": today_date
        }
    else:
        return {
            "from_date": frappe.datetime.week_start(),
            "to_date": frappe.datetime.week_end()
        }

def get_report_data(filters):
    task_filters = []
    
    if filters.get("project"):
        task_filters.append(["project", "=", filters.get("project")])
    
    if filters.get("status"):
        task_filters.append(["status", "=", filters.get("status")])
    
    if filters.get("priority"):
        task_filters.append(["priority", "=", filters.get("priority")])
    
    if filters.get("assigned_to"):
        task_filters.append(["_assign", "like", f'%"{filters.get("assigned_to")}"%'])
    
    # Time based filters for tasks
    if filters.get("filter_based_on") == "modified":
        task_filters.append(["modified", "between", [filters.get("from_date"), filters.get("to_date")]])
    elif filters.get("filter_based_on") == "creation":
        task_filters.append(["creation", "between", [filters.get("from_date"), filters.get("to_date")]])
    elif filters.get("filter_based_on") == "start_date":
        task_filters.append(["start_date", "between", [filters.get("from_date"), filters.get("to_date")]])
    elif filters.get("filter_based_on") == "end_date":
        task_filters.append(["end_date", "between", [filters.get("from_date"), filters.get("to_date")]])
    else:
        task_filters.append(["modified", "between", [filters.get("from_date"), filters.get("to_date")]])
    
    # Get all tasks matching the filters
    all_tasks = frappe.get_all(
        "Task",
        filters=task_filters,
        fields=[
            "name", "task_name", "project", "assigned_to", "_assign", 
            "start_date", "end_date", "status", "priority", "progress_",
            "creation", "modified", "time_in_minutes"
        ],
        order_by="modified desc"
    )
    
    # Mark backlog tasks if show_backlogs is enabled
    if filters.get("show_backlogs"):
        backlog_tasks = frappe.get_all(
            "Task",
            filters=[
                ["end_date", "<", today()],
                ["status", "not in", ["Completed", "Cancelled"]]
            ],
            fields=["name"],
            pluck="name"
        )
        
        backlog_set = set(backlog_tasks)
        for task in all_tasks:
            task["is_backlog"] = 1 if task["name"] in backlog_set else 0
    else:
        for task in all_tasks:
            task["is_backlog"] = 0
    
    today_date = getdate(today())
    
    for task in all_tasks:
        if task["status"] == "Completed":
            task["status_display"] = "🟢 Completed"
        elif task["status"] == "Cancelled":
            task["status_display"] = "⚫ Cancelled"
        elif task["status"] == "In Progress":
            task["status_display"] = "🟡 In Progress"
        elif task["status"] == "On Hold" or task["status"] == "Paused":
            task["status_display"] = "🟠 On Hold"
        elif task["is_backlog"]:
            task["status_display"] = "🔴 Backlog"
        else:
            task["status_display"] = "⚪ " + task["status"]
        
        if task["end_date"]:
            days_left = date_diff(task["end_date"], today_date)
            if days_left < 0:
                task["days_left"] = f"🔴 {abs(days_left)} days overdue"
            elif days_left == 0:
                task["days_left"] = "⚠️ Due today"
            else:
                task["days_left"] = f"🟢 {days_left} days left"
        else:
            task["days_left"] = ""
        
        task["deliverable"] = get_deliverable_for_task(task["name"])
        
        assigned_users = []
        if task["_assign"]:
            try:
                assigned_data = json.loads(task["_assign"])
                for user_id in assigned_data:
                    user_name = get_user_name(user_id)
                    assigned_users.append(user_name)
            except:
                pass
        
        if not assigned_users and task["assigned_to"]:
            user_name = get_user_name(task["assigned_to"])
            assigned_users.append(user_name)
        
        task["assigned_users"] = ", ".join(assigned_users) if assigned_users else ""
        
        if task["time_in_minutes"]:
            hours = int(task["time_in_minutes"] // 60)
            minutes = int(task["time_in_minutes"] % 60)
            task["time_spent"] = f"{hours}h {minutes}m"
        else:
            task["time_spent"] = ""
    
    project_filters = []
    
    if filters.get("project"):
        project_filters.append(["name", "=", filters.get("project")])
    
    projects = frappe.get_all(
        "Project",
        filters=project_filters,
        fields=[
            "name", "project_name", "status", "priority", 
            "start_date", "end_date", "progress", "modified"
        ],
        order_by="modified desc"
    )
    
    project_map = {p.name: p for p in projects}
    
    for task in all_tasks:
        if task["project"] in project_map:
            project = project_map[task["project"]]
            task["project_name"] = project.project_name
            task["project_status"] = project.status
            task["project_progress"] = project.progress
        else:
            task["project_name"] = task["project"]
            task["project_status"] = ""
            task["project_progress"] = 0
    
    return all_tasks

def get_deliverable_for_task(task_id):
    deliverable_task = frappe.get_all(
        "Deliverable Task",
        filters={"task": task_id},
        fields=["parent"],
        limit=1
    )
    
    if deliverable_task:
        deliverable = frappe.get_all(
            "Deliverable",
            filters={"name": deliverable_task[0].parent},
            fields=["name", "deliverable_name", "status"],
            limit=1
        )
        
        if deliverable:
            return f"{deliverable[0].deliverable_name} ({deliverable[0].status})"
    
    return ""

def get_user_name(user_id):
    if not user_id:
        return ""
    
    user_name = frappe.cache().hget("user_fullname", user_id)
    
    if not user_name:
        user = frappe.get_all(
            "User",
            filters={"name": user_id},
            fields=["full_name"],
            limit=1
        )
        
        if user:
            user_name = user[0].full_name
            frappe.cache().hset("user_fullname", user_id, user_name)
        else:
            user_name = user_id
    
    return user_name

def get_columns(filters):
    return [
        {"label": "Task", "fieldname": "task_name", "fieldtype": "Data", "width": 200},
        {"label": "Project", "fieldname": "project", "fieldtype": "Link", "options": "Project", "width": 120},
        {"label": "Deliverable", "fieldname": "deliverable", "fieldtype": "Data", "width": 180},
        {"label": "Status", "fieldname": "status_display", "fieldtype": "Data", "width": 120},
        {"label": "Priority", "fieldname": "priority", "fieldtype": "Data", "width": 80},
        {"label": "Assigned To", "fieldname": "assigned_users", "fieldtype": "Data", "width": 150},
        {"label": "Progress", "fieldname": "progress_", "fieldtype": "Percent", "width": 100},
        {"label": "Timeline", "fieldname": "days_left", "fieldtype": "Data", "width": 130},
        {"label": "Start Date", "fieldname": "start_date", "fieldtype": "Date", "width": 100},
        {"label": "Due Date", "fieldname": "end_date", "fieldtype": "Date", "width": 100},
        {"label": "Time Spent", "fieldname": "time_spent", "fieldtype": "Data", "width": 100},
        {"label": "Last Modified", "fieldname": "modified", "fieldtype": "Datetime", "width": 150},
    ]

def get_task_status_chart(data):
    if not data:
        return None
    
    status_counts = {}
    for task in data:
        status = task["status"]
        if status not in status_counts:
            status_counts[status] = 0
        status_counts[status] += 1
    
    return {
        "data": {
            "labels": list(status_counts.keys()),
            "datasets": [{"name": "Task Status", "values": list(status_counts.values())}]
        },
        "type": "pie",
        "height": 300,  
        "colors": ["#4caf50", "#ff5722", "#2196f3", "#ff9800", "#9c27b0", "#607d8b"],
        "title": "Task Status Distribution"
    }

def get_report_summary(data, filters):
    if not data:
        return []
    
    total_tasks = len(data)
    completed_tasks = sum(1 for task in data if task["status"] == "Completed")
    in_progress_tasks = sum(1 for task in data if task["status"] == "In Progress")
    backlog_tasks = sum(1 for task in data if task["is_backlog"])
    
    total_minutes = sum(task["time_in_minutes"] or 0 for task in data)
    total_hours = total_minutes / 60
    
    completion_rate = (completed_tasks / total_tasks * 100) if total_tasks > 0 else 0
    
    progress_values = [task["progress_"] for task in data if task["progress_"] is not None]
    avg_progress = sum(progress_values) / len(progress_values) if progress_values else 0
    
    date_range = f"{filters.get('from_date')} to {filters.get('to_date')}"
    
    return [
        {"label": "Time Period", "value": date_range, "indicator": "blue"},
        {"label": "Total Tasks", "value": total_tasks, "indicator": "blue"},
        {"label": "Completed", "value": completed_tasks, "indicator": "green"},
        {"label": "In Progress", "value": in_progress_tasks, "indicator": "orange"},
        {"label": "Backlog", "value": backlog_tasks, "indicator": "red"},
        {"label": "Completion Rate", "value": f"{completion_rate:.1f}%", "indicator": "green" if completion_rate > 50 else "orange"},
        {"label": "Avg Progress", "value": f"{avg_progress:.1f}%", "indicator": "blue"},
        {"label": "Total Hours", "value": f"{total_hours:.1f}h", "indicator": "purple"}
    ]

File: project_management/project_management/report/task_progress_timeline/task_progress_timeline.js
// Copyright (c) 2025, Aman and Yashvi and contributors
// For license information, please see license.txt

frappe.query_reports["Task Progress Timeline"] = {
	"filters": [
        {
            "fieldname": "project",
            "label": __("Project"),
            "fieldtype": "Link",
            "options": "Project",
            "width": "100px"
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nPlanned\nIn Progress\nOn Hold\nPaused\nCompleted\nCancelled",
            "width": "100px"
        },
        {
            "fieldname": "priority",
            "label": __("Priority"),
            "fieldtype": "Select",
            "options": "\nLow\nMedium\nHigh\nCritical",
            "width": "100px"
        },
        {
            "fieldname": "assigned_to",
            "label": __("Assigned To"),
            "fieldtype": "Link",
            "options": "User",
            "width": "100px"
        },
        {
            "fieldname": "filter_based_on",
            "label": __("Filter Based On"),
            "fieldtype": "Select",
            "options": "modified|Last Modified\ncreation|Creation Date\nstart_date|Start Date\nend_date|Due Date",
            "default": "modified",
            "width": "100px"
        },
        {
            "fieldname": "time_interval",
            "label": __("Time Interval"),
            "fieldtype": "Select",
            "options": "\ntoday\nyesterday\nthis_week\nlast_week\nthis_month\nlast_month\nthis_quarter\nlast_quarter\nthis_year\nlast_year\nall",
            "default": "this_week",
            "width": "100px"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "width": "80px",
            "depends_on": "eval:doc.time_interval==''"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "width": "80px",
            "depends_on": "eval:doc.time_interval==''"
        },
        {
            "fieldname": "show_backlogs",
            "label": __("Show Backlogs"),
            "fieldtype": "Check",
            "default": 1
        }
    ],
    
    "onload": function(report) {
        update_date_range(report);
        
        report.page.fields_dict.time_interval.$input.on("change", function() {
            update_date_range(report);
            report.refresh();
        });
        
        report.page.add_inner_button(__('Reset Filters'), function() {
            report.clear_filter_values();
            report.set_filter_value('time_interval', 'this_week');
            update_date_range(report);
            report.refresh();
        });
        
        report.page.add_inner_button(__('Today'), function() {
            report.set_filter_value('time_interval', 'today');
            update_date_range(report);
            report.refresh();
        }, __("Quick Filters"));
        
        report.page.add_inner_button(__('This Week'), function() {
            report.set_filter_value('time_interval', 'this_week');
            update_date_range(report);
            report.refresh();
        }, __("Quick Filters"));
        
        report.page.add_inner_button(__('Last Week'), function() {
            report.set_filter_value('time_interval', 'last_week');
            update_date_range(report);
            report.refresh();
        }, __("Quick Filters"));
        
        report.page.add_inner_button(__('This Month'), function() {
            report.set_filter_value('time_interval', 'this_month');
            update_date_range(report);
            report.refresh();
        }, __("Quick Filters"));
        
        report.page.add_inner_button(__('Export PDF'), function() {
            frappe.call({
                method: "frappe.desk.reportview.export_query",
                args: {
                    title: "Daily Weekly Progress Report",
                    filters: report.get_values(),
                    report_name: "Daily Weekly Progress Report",
                    file_format_type: "PDF"
                },
                callback: function(r) {
                    if (r.message) {
                        window.open('/api/method/frappe.utils.print_format.download_pdf?file_url=' + encodeURIComponent(r.message));
                    }
                }
            });
        }, __("Export"));
        
        report.page.add_inner_button(__('Export Excel'), function() {
            frappe.call({
                method: "frappe.desk.reportview.export_query",
                args: {
                    title: "Daily Weekly Progress Report",
                    filters: report.get_values(),
                    report_name: "Daily Weekly Progress Report",
                    file_format_type: "Excel"
                },
                callback: function(r) {
                    if (r.message) {
                        window.location.href = '/api/method/frappe.utils.file_manager.download_file?file_url=' + encodeURIComponent(r.message);
                    }
                }
            });
        }, __("Export"));
    },
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (!data) return value;
        
        if (column.fieldname == "progress_" && data.progress_ !== undefined && data.progress_ !== null) {
            let progress = data.progress_;
            if (progress < 30) {
                value = `<span style="color: #ff5858">${progress}%</span>`;
            } else if (progress < 70) {
                value = `<span style="color: #ffa00a">${progress}%</span>`;
            } else {
                value = `<span style="color: #28a745">${progress}%</span>`;
            }
        }
        
        if (column.fieldname == "days_left" && data.days_left) {
            if (data.days_left.includes("overdue")) {
                value = `<span style="color: #ff5858; font-weight: bold;">${data.days_left}</span>`;
            } else if (data.days_left.includes("today")) {
                value = `<span style="color: #ffa00a; font-weight: bold;">${data.days_left}</span>`;
            } else {
                value = `<span style="color: #28a745">${data.days_left}</span>`;
            }
        }
        
        if (column.fieldname == "status_display" && data.status_display) {
            if (data.status_display.includes("🟢")) {
                value = `<span style="color: #28a745; font-weight: bold;">${data.status_display}</span>`;
            } else if (data.status_display.includes("🟡")) {
                value = `<span style="color: #ffa00a; font-weight: bold;">${data.status_display}</span>`;
            } else if (data.status_display.includes("🔴")) {
                value = `<span style="color: #ff5858; font-weight: bold;">${data.status_display}</span>`;
            } else if (data.status_display.includes("⚫")) {
                value = `<span style="color: #6c757d; font-weight: bold;">${data.status_display}</span>`;
            } else if (data.status_display.includes("🟠")) {
                value = `<span style="color: #fd7e14; font-weight: bold;">${data.status_display}</span>`;
            } else {
                value = `<span style="color: #6c757d;">${data.status_display}</span>`;
            }
        }
        
        if (column.fieldname == "priority" && data.priority) {
            let priority = data.priority;
            let color = "";
            
            if (priority === "Critical") {
                color = "#ff5858";
            } else if (priority === "High") {
                color = "#ffa00a";
            } else if (priority === "Medium") {
                color = "#5e64ff";
            } else {
                color = "#28a745";
            }
            
            value = `<span style="color: ${color}; font-weight: bold;">${priority}</span>`;
        }
        
        if (column.fieldname == "task_name") {
            if (data.is_backlog === 1) {
                value = `<div style="background-color: rgba(255, 88, 88, 0.1); padding: 2px 5px; border-radius: 3px;">
                    <span style="color: #ff5858; font-weight: bold;">⚠️ </span>${value}
                </div>`;
            } else if (data.end_date && frappe.datetime.get_diff(data.end_date, frappe.datetime.get_today()) < 0 && 
                      data.status !== "Completed" && data.status !== "Cancelled") {
                value = `<div style="background-color: rgba(255, 88, 88, 0.1); padding: 2px 5px; border-radius: 3px;">
                    <span style="color: #ff5858; font-weight: bold;">⚠️ </span>${value}
                </div>`;
            }
        }

        if (column.fieldname == "time_spent" && data.time_spent) {
            value = `<span style="color: #7a3db8;">${data.time_spent}</span>`;
        }
        
        return value;
    }
};

function update_date_range(report) {
    var time_interval = report.get_filter_value('time_interval') || 'this_week';
    var today = frappe.datetime.get_today();
    var from_date, to_date;
    
    switch(time_interval) {
        case 'today':
            from_date = today;
            to_date = today;
            break;
        case 'yesterday':
            from_date = frappe.datetime.add_days(today, -1);
            to_date = frappe.datetime.add_days(today, -1);
            break;
        case 'this_week':
            from_date = frappe.datetime.week_start();
            to_date = frappe.datetime.week_end();
            break;
        case 'last_week':
            from_date = frappe.datetime.add_days(frappe.datetime.week_start(), -7);
            to_date = frappe.datetime.add_days(frappe.datetime.week_end(), -7);
            break;
        case 'this_month':
            from_date = frappe.datetime.month_start();
            to_date = frappe.datetime.month_end();
            break;
        case 'last_month':
            from_date = frappe.datetime.add_months(frappe.datetime.month_start(), -1);
            to_date = frappe.datetime.add_months(frappe.datetime.month_end(), -1);
            break;
        case 'this_quarter':
            from_date = frappe.datetime.quarter_start();
            to_date = frappe.datetime.quarter_end();
            break;
        case 'last_quarter':
            from_date = frappe.datetime.add_months(frappe.datetime.quarter_start(), -3);
            to_date = frappe.datetime.add_months(frappe.datetime.quarter_end(), -3);
            break;
        case 'this_year':
            from_date = frappe.datetime.year_start();
            to_date = frappe.datetime.year_end();
            break;
        case 'last_year':
            from_date = frappe.datetime.add_years(frappe.datetime.year_start(), -1);
            to_date = frappe.datetime.add_years(frappe.datetime.year_end(), -1);
            break;
        case 'all':
            from_date = '2000-01-01';
            to_date = frappe.datetime.get_today();
            break;
        default:
            // Default to this week if not specified
            from_date = frappe.datetime.week_start();
            to_date = frappe.datetime.week_end();
    }
    
    report.set_filter_value('from_date', from_date);
    report.set_filter_value('to_date', to_date);
}
